### PR TITLE
ISPN-13504 Implement cache sets using publishers

### PR DESCRIPTION
**Pull request by danberindei**
https://issues.redhat.com/browse/ISPN-13504

~~Depends on #9677~~

Implement only localPublisher methods in internal cache sets
Make iteration and streams go through DistributedCacheStream
even with CACHE_MODE_LOCAL
Make EntrySetCommand and KeySetCommand return an InternalCacheSet
that only implements localPublisher()
Pass all flags and lock owner to cache operations
Delete DistributionBulkInterceptor
Handle tx context entries in LocalClusterPublisherManagerImpl only
Unwrap CompletionException in publisher responses
Remove unused cache sets

### CODE REVIEW

**Comment by wburns** [comment]
Changes overall look good to me in general. I won't be able to do a full review until next week though.

File: core/src/main/java/org/infinispan/commands/CommandsFactory.java [hunk 1]
```diff
@@ -613,17 +613,17 @@ default SingleRpcCommand buildSingleRpcCommand(ReplicableCommand call) {
    BackupNoopCommand buildBackupNoopCommand();
 
    <K, R> ReductionPublisherRequestCommand<K> buildKeyReductionPublisherCommand(boolean parallelStream, DeliveryGuarantee deliveryGuarantee,
-         IntSet segments, Set<K> keys, Set<K> excludedKeys, boolean includeLoader,
+         IntSet segments, Set<K> keys, Set<K> excludedKeys, long explicitFlags,
```
Comment (wburns) [inline core/src/main/java/org/infinispan/commands/CommandsFactory.java:616]: What flags do you envision we support now?

Comment (wburns) [reply core/src/main/java/org/infinispan/commands/CommandsFactory.java:616]: We probably should document which ones do anything on the Manager interface so we are aware. It was pretty simple to tell we only supported turning off loader or not before.

File: core/src/main/java/org/infinispan/interceptors/impl/CacheLoaderInterceptor.java [hunk 2]
```diff
@@ -265,8 +241,8 @@ public Object visitKeySetCommand(InvocationContext ctx, KeySetCommand command)
             return rv;
          }
 
-         CacheSet<K> keySet = (CacheSet<K>) rv;
```
Comment (wburns) [inline core/src/main/java/org/infinispan/interceptors/impl/CacheLoaderInterceptor.java:268]: You can remove the REMOTE_ITERATION optimization above for the KeySetCommand as well and the value in the FlagBitSets now that we don't use an iterator.

File: core/src/main/java/org/infinispan/reactive/publisher/impl/ClusterPublisherManagerImpl.java [hunk 3]
```diff
@@ -1042,8 +1024,7 @@ void completeSegment(int segment) {
        * via {@link Subscriber#onError(Throwable)}.
        */
       boolean handleThrowable(Throwable t, Address target, IntSet segments) {
-         // Most likely SuspectException will be wrapped in CompletionException
-         if (t instanceof SuspectException || t.getCause() instanceof SuspectException) {
+         if (t instanceof SuspectException) {
```
Comment (wburns) [inline core/src/main/java/org/infinispan/reactive/publisher/impl/ClusterPublisherManagerImpl.java:1027]: Did you find the only wrapping is a CompletionException? Which I see is handled at `InnerPublisherSubscription#handleThrowInResponse` invocation.

File: core/src/main/java/org/infinispan/reactive/publisher/impl/LocalPublisherManagerImpl.java [hunk 4]
```diff
@@ -83,12 +94,6 @@
    protected int maxSegment;
    protected final int cpuCount = ProcessorInfo.availableProcessors();
 
-   protected CacheSet<K> keySet;
```
Comment (wburns) [inline core/src/main/java/org/infinispan/reactive/publisher/impl/LocalPublisherManagerImpl.java:86]: I wonder if not caching these has caused a performance regression. I know originally I found going down the interceptor stack for every invocation had a measurable cost.

File: core/src/main/java/org/infinispan/reactive/publisher/impl/commands/batch/NextPublisherCommand.java [hunk 5]
```diff
@@ -14,8 +14,6 @@
 public class  NextPublisherCommand extends BaseRpcCommand implements TopologyAffectedCommand {
    public static final byte COMMAND_ID = 25;
 
-   private PublisherHandler publisherHandler;
```
Comment (wburns) [inline core/src/main/java/org/infinispan/reactive/publisher/impl/commands/batch/NextPublisherCommand.java:17]: Whoops :)

**Review by wburns — COMMENTED** [review]
Looks pretty darn good to me, a lot of no longer needed code removed 🎉

File: core/src/main/java/org/infinispan/CacheCollection.java [hunk 6]
```diff
@@ -19,36 +13,4 @@
...
-    * to the original Cache used to create this CacheCollection.
-    * @param segment the segment that all published elements belong to
-    * @return Publisher that will publish the elements for the given segment
-    * @implSpec Default implementation just does:
-    * <pre> {@code
-    * return localPublisher(org.infinispan.commons.util.IntSets.immutableSet(segment));
-    * }</pre>
-    */
-   @Experimental
-   default Publisher<E> localPublisher(int segment) {
```
[the remark this answers is not in the document]
Comment (danberindei) [reply core/src/main/java/org/infinispan/CacheCollection.java:35]: I wanted to move them so that I could make changes to them.
In the end I didn't make any change, so I could put them back.

File: anchored-keys/src/main/java/org/infinispan/anchored/impl/AnchoredFetchInterceptor.java [hunk 7]
```diff
@@ -267,75 +264,37 @@ public FetchResponseCollector(K key) {
...
-         }
-         CacheEntry<K, V> localEntry = iterator.next();
+      private CacheEntry<K, V> fixPublisherEntry(int segment, CacheEntry<K, V> localEntry) {
          if (localEntry.getMetadata() instanceof RemoteMetadata) {
             // The key exists and the anchor location is a remote node
             RemoteMetadata remoteMetadata = (RemoteMetadata) localEntry.getMetadata();
             Address keyLocation = remoteMetadata.getAddress();
 
-            int segment = distributionManager.getCacheTopology().getSegment(localEntry.getKey());
             return CompletionStages.join(getRemoteValue(keyLocation, localEntry.getKey(), segment, false));
```
[the remark this answers is not in the document]
Comment (danberindei) [reply anchored-keys/src/main/java/org/infinispan/anchored/impl/AnchoredFetchInterceptor.java:295]: I'm not sure I want to do that, because then I'd have to also worry about what's a good number of concurrent requests -- I don't trust flatMap's defaults ;)

Thanks for noticing the code is blocking though, looks like I moved from iterator to publisher without actually reading the implementation!

[the remark this answers is not in the document]
Comment (wburns) [reply anchored-keys/src/main/java/org/infinispan/anchored/impl/AnchoredFetchInterceptor.java:295]: Well it isn't necessarily faster, but the code don't have to check the segment or allocate a Maybe for each of the local elements.

Either way is fine though, I don't think it is a big deal. But it does need to use `concatMap` at a minimum to handle the blocking part.

File: core/src/main/java/org/infinispan/cache/impl/AbstractDelegatingAdvancedCache.java [hunk 8]
```diff
@@ -593,14 +591,6 @@ public DataConversion getValueDataConversion() {
       return cache.getValueDataConversion();
    }
 
-   protected final void putForExternalRead(K key, V value, EnumSet<Flag> flags, ClassLoader classLoader) {
```
[the remark this answers is not in the document]
Comment (danberindei) [reply core/src/main/java/org/infinispan/cache/impl/AbstractDelegatingAdvancedCache.java:596]: I did mean to include it in the PR, Galder removed the `withClassloader` stuff back in 2016 and these just stuck around but they're not used/usable any more.
I can move them to a separate commit if it's too confusing.

File: core/src/main/java/org/infinispan/cache/impl/CacheBackedEntrySet.java [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+
+   @Override
+   public boolean contains(Object o) {
+      if (!(o instanceof Map.Entry))
+         return false;
+
+      Map.Entry<?, ?> entry = (Map.Entry<?, ?>) o;
+
+      InvocationContext ctx = cache.invocationContextFactory.createInvocationContext(false, 1);
+      V cacheValue = cache.get(entry.getKey(), explicitFlags, ctx);
```
[the remark this answers is not in the document]
Comment (danberindei) [reply core/src/main/java/org/infinispan/cache/impl/CacheBackedEntrySet.java:32]: Actually `CacheBackedKeySet.contains()` is wrong, it should be using the explicit flags (the cache field is a `CacheImpl` without any flags).

File: core/src/main/java/org/infinispan/commands/CommandsFactory.java (hunk 1 above)
Comment (danberindei) [reply core/src/main/java/org/infinispan/commands/CommandsFactory.java:616]: I'm not sure it's useful: the way I see it, ClusterPublisherManager just passes the flags along to the KeySetCommand/CacheSetCommand, and we support whatever flags the interceptors decide to support.
If someone decides an interceptor should support a new flag, they're unlikely to remember to update the ClusterPublisherManager javadoc.

File: core/src/main/java/org/infinispan/reactive/publisher/impl/ClusterPublisherManagerImpl.java (hunk 3 above)
Comment (danberindei) [reply core/src/main/java/org/infinispan/reactive/publisher/impl/ClusterPublisherManagerImpl.java:1027]: Yes, there's nothing else wrapping SuspectException

File: core/src/main/java/org/infinispan/reactive/publisher/impl/LocalPublisherManagerImpl.java (hunk 4 above)
Comment (danberindei) [reply core/src/main/java/org/infinispan/reactive/publisher/impl/LocalPublisherManagerImpl.java:86]: The same thing happens when you call the `localPublisher()` method: every cache set wrapper that an interceptor added now wraps the publisher (usually more than once). Plus I removed 1 interceptor, so it should be faster now ;)

File: core/src/main/java/org/infinispan/reactive/publisher/impl/commands/batch/NextPublisherCommand.java (hunk 5 above)
Comment (danberindei) [reply core/src/main/java/org/infinispan/reactive/publisher/impl/commands/batch/NextPublisherCommand.java:17]: It wasn't breaking anything, just sitting there :)

**Comment by danberindei** [comment]
Updated to address review comments and deprecate the REMOTE_ITERATION flag.

**Review by wburns — COMMENTED** [review]
Just some minor stuff

File: anchored-keys/src/main/java/org/infinispan/anchored/impl/AnchoredFetchInterceptor.java [hunk 10]
```diff
@@ -267,75 +266,42 @@ public FetchResponseCollector(K key) {
...
-      }
-
-      @Override
-      public void forEachRemaining(Consumer<? super CacheEntry<K, V>> action) {
+      private Single<CacheEntry<K, V>> fixPublisherEntry(int segment, CacheEntry<K, V> localEntry) {
+         // The key exists and the anchor location is a remote node
+         RemoteMetadata remoteMetadata = (RemoteMetadata) localEntry.getMetadata();
+         Address keyLocation = remoteMetadata.getAddress();
 
+         return Single.fromCompletionStage(getRemoteValue(keyLocation, localEntry.getKey(), segment, false));
```
[the remark this answers is not in the document]
Comment (danberindei) [reply anchored-keys/src/main/java/org/infinispan/anchored/impl/AnchoredFetchInterceptor.java:304]: Oops, you're right. It could be removed on the "anchor" owner but not yet on the node doing the iteration.

File: core/src/main/java/org/infinispan/cache/impl/CacheBackedEntrySet.java [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+   protected Object extractKey(Object e) {
+      if (!(e instanceof Map.Entry))
+         return null;
+
+      return ((Map.Entry<?, ?>) e).getKey();
+   }
+
+   @Override
+   protected CacheEntry<K, V> wrapElement(CacheEntry<K, V> e) {
+      return new ForwardingCacheEntry<K, V>() {
```
[the remark this answers is not in the document]
Comment (danberindei) [reply core/src/main/java/org/infinispan/cache/impl/CacheBackedEntrySet.java:52]: REMOTE_ITERATION wouldn't be set on the origin node, so yeah it was the same before.

But I see what you mean, `forEach()` and `removeIf` could bypass the operator and avoid wrapping each entry because they don't need `setValue()`.

[the remark this answers is not in the document]
Comment (danberindei) [reply core/src/main/java/org/infinispan/cache/impl/CacheBackedEntrySet.java:52]: Done

**Comment by danberindei** [comment]
Squashed

File: anchored-keys/src/main/java/org/infinispan/anchored/impl/AnchoredFetchInterceptor.java [hunk 12]
```diff
@@ -267,75 +266,42 @@ public FetchResponseCollector(K key) {
...
-      public void remove() {
-
+      private Flowable<CacheEntry<K, V>> fixPublisher(Publisher<CacheEntry<K, V>> nextPublisher, Integer segment) {
+         return Flowable.fromPublisher(nextPublisher)
+                        .groupBy(entry -> entry.getMetadata() instanceof RemoteMetadata)
+                        .flatMap(gf -> {
+                           if (!gf.getKey()) {
+                              return gf;
+                           }
+                           return gf.concatMapMaybe(incompleteLocalEntry -> {
```
Comment (wburns) [inline anchored-keys/src/main/java/org/infinispan/anchored/impl/AnchoredFetchInterceptor.java:293]: Guessing the spacing is off again for the brackets to be needed? :D

**Review by wburns — APPROVED** [review]
LGTM, just waiting on CI.

**Comment by wburns** [comment]
Oops I didn't even notice there are two failures that have been failing the entire time :( https://ci.infinispan.org/job/Infinispan/job/PR-9678/10/testReport/junit/org.infinispan.multimap.impl/TxEmbeddedMultimapCacheTest/testSizeInExplicitTxWithRemoveWithPredicate/

File: anchored-keys/src/main/java/org/infinispan/anchored/impl/AnchoredFetchInterceptor.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (danberindei) [reply anchored-keys/src/main/java/org/infinispan/anchored/impl/AnchoredFetchInterceptor.java]: Yep, I did it with flatMap in the end.

File: anchored-keys/src/main/java/org/infinispan/anchored/impl/AnchoredFetchInterceptor.java (hunk 12 above)
Comment (danberindei) [reply anchored-keys/src/main/java/org/infinispan/anchored/impl/AnchoredFetchInterceptor.java:293]: The parameter name was too long... I renamed to `incompleteEntry` and now it fits on one line :)

**Comment by danberindei** [comment]
Fixed @wburns 

`EmbeddedMultiMapCache.remove()` was using `cache.keySet().stream().forEach((cache1, key) -> cache1.computeIfPresent(key, ...))` to iterate over keys, and to modify the buckets containing the value.
In local caches, the stream would invoke the `forEach()` lambda in the same thread, so all the modifications would join the running transaction.

With the publisher-based streams, the `forEach()` lambda runs on a blocking thread, so the modifications don't join the running transaction. Most important, in a distributed cache the stream `forEach()` lambda would run on different nodes, so there's no way it could join the caller thread's transaction. So I modificed `EmbeddedMultiMapCache.remove()` to iterate over the keys in the same thread, with `cache.keySet().forEach()`.

I took the opportunity to go over the cache set operations that perform writes (Entry.setValue(), Iterator.remove(), Set.removeIf()) and make them correctly pass flags + lock owner to the cache operation.

File: multimap/src/main/java/org/infinispan/multimap/impl/EmbeddedMultimapCache.java [hunk 13]
```diff
@@ -188,7 +189,7 @@ private Boolean containsEntryInternal(V value) {
    }
 
    private Long sizeInternal() {
-      return cache.values().parallelStream().mapToLong(Bucket::size).sum();
+      return cache.entrySet().parallelStream().mapToLong(entry -> entry.getValue().size()).sum();
```
Comment (wburns) [inline multimap/src/main/java/org/infinispan/multimap/impl/EmbeddedMultimapCache.java:192]: Is this change really needed? I would expect values to work properly.

Comment (danberindei) [reply multimap/src/main/java/org/infinispan/multimap/impl/EmbeddedMultimapCache.java:192]: It's not really needed, I was just thinking it's less work if we have a single "map" operation instead of mapping to the value (done inside `ValuesCacheCollection.stream()`) and then mapping the value to the bucket size.

**Comment by wburns** [comment]
Integrated into main, thanks @danberindei !
